src/m3ninx/index/segment/builder/builder.go

```diff
@@ -41,20 +43,131 @@ var (
 
 const (
 	// Slightly buffer the work to avoid blocking main thread.
-	indexQueueSize = 2 << 9 // 1024
+	indexQueueSize     = 2 << 9 // 1024
+	entriesPerIndexJob = 32
 )
 
+var (
+	globalIndexWorkers  = &indexWorkers{}
+	fieldsMapSetOptions = fieldsMapSetUnsafeOptions{
+		// Builder takes ownership of keys and docs so it's ok
+		// to avoid copying and finalizing keys.
+		NoCopyKey:     true,
+		NoFinalizeKey: true,
+	}
+)
+
+type indexWorkers struct {
+	sync.RWMutex
+	builders int
+	queues   []chan indexJob
+}
+
 type indexJob struct {
 	wg *sync.WaitGroup
 
-	id    postings.ID
-	field doc.Field
+	opts Options
+
+	entries     [entriesPerIndexJob]indexJobEntry
+	usedEntries int
+
+	shard         int
+	shardedFields *shardedFields
 
-	shard    int
-	idx      int
 	batchErr *index.BatchPartialError
 }
 
+type indexJobEntry struct {
+	id     postings.ID
+	field  doc.Field
+	docIdx int
+}
+
+func (w *indexWorkers) registerBuilder() {
+	w.Lock()
+	defer w.Unlock()
+
+	preIncBuilders := w.builders
+	w.builders++
+
+	if preIncBuilders != 0 {
+		return // Already initialized.
+	}
+
+	// Need to initialize structures, prepare all num CPU
+	// worker queues, even if we don't use all of them.
+	n := runtime.NumCPU()
```

Comment:
This won't cause the same issues where we create numcpu workers for each namespace will it?

Reply:
Thankfully this datastructure is global to this package:
```
globalIndexWorkers  = &indexWorkers{}
```

So there is only one single instance of Num CPU workers globally that is shared by any builder doing work now 👍